### PR TITLE
Firebase terraform

### PR DESCRIPTION
**Pull request by binnmti**
## Release Notes

- **New Features**
  - Introduced a tag-based deployment process for streamlined production updates.
  - Enhanced staging deployment workflow to utilize staging-specific credentials and environment variables.
  - Added documentation for managing Google and GitHub authentication client secrets with Terraform.

- **Bug Fixes**
  - Temporarily commented out the test execution step in the integration testing workflow due to an unspecified error.

- **Chores**
  - Updated deployment workflow to trigger on pull requests instead of direct pushes.
  - Improved Firebase project setup with comprehensive Terraform configurations for better management and security.
  - Provided guidelines for updating the staging environment and managing pricing plans in Terraform documentation.

### CODE REVIEW

File: firebase-terraform/firestore.rules [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+                      || request.resource.data.imageUrl is string
+                      && request.resource.data.imageUrl.matches('https?://.*'));
+
+      allow delete: if false;
+    }
+
+    match /fcmTokens/{token} {
+      allow read: if request.auth != null;
+      allow write: if request.auth != null;
+    }
```
[the remark this answers is not in the document]
Comment (binnmti) [reply firebase-terraform/firestore.rules]: This content has been changed so please ignore it.

File: firebase-terraform/firestore.rules [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+
+    match /{document=**} {
+      allow read, write: if request.auth != null;
+    }
```
[the remark this answers is not in the document]
Comment (binnmti) [reply firebase-terraform/firestore.rules]: This content has been changed so please ignore it.

File: .github/workflows/deploy-staging.yml [hunk 3]
```diff
@@ -1,19 +1,20 @@
 name: deploy
 
 on:
-  push:
+  pull_request:
     branches:
       - main
 
+permissions: write-all
+
```
[the remark this answers is not in the document]
Comment (binnmti) [reply .github/workflows/deploy-staging.yml]: I got an error when I wrote the following. How should I write it?

```
permissions:
  deployments: write
```

File: one-pager-maker/firebase.json [hunk 4]
```diff
@@ -1,25 +1,12 @@
 {
   "hosting": {
-    "public": "one-pager-maker/dist",
+    "public": "public",
```
[the remark this answers is not in the document]
Comment (binnmti) [reply one-pager-maker/firebase.json]: revert

File: firebase-terraform/firestore.rules [hunk 5]
```diff
@@ -0,0 +1,9 @@
+service cloud.firestore { 
+  match /databases/{database}/documents {
+    match /{documents=**} {
+      allow read;
+      allow create, update: if request.auth != null;
+      allow delete: if false;
+    }
```
[the remark this answers is not in the document]
Comment (binnmti) [reply firebase-terraform/firestore.rules:7]: If you change it to what you pointed out, you will get the following error:

```userServiceImpl.ts:64  FirebaseError: Missing or insufficient permissions.```

File: firebase-terraform/main.tf [hunk 6]
```diff
@@ -0,0 +1,286 @@
...
+  sensitive   = true
+}
+
+# githubのclient_id to client_secret
+resource "google_identity_platform_default_supported_idp_config" "github_sign_in" {
+  provider      = google-beta
+  project       = google_firebase_project.default.project
+  enabled       = true
+  idp_id        = "github.com"
+  client_id     = "Ov23lirVDPHp9Vx8xeOy"
```
[the remark this answers is not in the document]
Comment (binnmti) [reply firebase-terraform/main.tf:141]: fixed

File: .github/workflows/deploy-staging.yml (hunk 3 above)
[the remark this answers is not in the document]
Comment (binnmti) [reply .github/workflows/deploy-staging.yml]: fixed

File: firebase-terraform/firestore.rules (hunk 5 above)
[the remark this answers is not in the document]
Comment (binnmti) [reply firebase-terraform/firestore.rules:7]: I don't know the correct settings, so I'll post this as a separate PR. Sorry.

File: one-pager-maker/public/404.html
Comment (binnmti) [inline one-pager-maker/public/404.html:1]: firebase.jsonでpublicではなくdistを見ているからで、デプロイしても問題はなかったので消した

File: one-pager-maker/firestore.rules
Comment (binnmti) [inline one-pager-maker/firestore.rules:1]: このルールは結果的には反映していなかった。terraform側に移動した

File: firebase.json [hunk 7]
```diff
@@ -13,13 +13,6 @@
       }
     ]
   },
-  "firestore": {
```
Comment (binnmti) [inline firebase.json:16]: terraform側に移動

File: .github/workflows/deploy-staging.yml
Comment (binnmti) [inline .github/workflows/deploy-staging.yml:1]: mainにマージされたタイミングはステージングにデプロイするようにした。(本番環境へのデプロイは別のPRでgitlab-flowのproductionで設定予定)

File: .github/workflows/integration-test.yml [hunk 8]
```diff
@@ -27,15 +27,17 @@ jobs:
...
-        run: bun run test-integration
+          echo "VITE_API_KEY=${{ secrets.STAGING_VITE_API_KEY }}" >> .env
+          echo "VITE_AUTH_DOMAIN=${{ secrets.STAGING_VITE_AUTH_DOMAIN }}" >> .env
+          echo "VITE_PROJECT_ID=${{ secrets.STAGING_VITE_PROJECT_ID }}" >> .env
+          echo "VITE_STORAGE_BUCKET=${{ secrets.STAGING_VITE_STORAGE_BUCKET }}" >> .env
+          echo "VITE_MESSAGING_SENDER_ID=${{ secrets.STAGING_VITE_MESSAGING_SENDER_ID }}" >> .env
+          echo "VITE_APP_ID=${{ secrets.STAGING_VITE_APP_ID }}" >> .env
+          echo "VITE_MEASUREMENT_ID=${{ secrets.STAGING_VITE_MEASUREMENT_ID }}" >> .env
+
+      # I don't know the cause of the error at the moment, so I'll stop it.
```
Comment (binnmti) [inline .github/workflows/integration-test.yml:38]: 現状エラーなので止めた

**Review by Procrustes5 — APPROVED** [review]
LGTM
ちなみにですが、mdファイルは英語化はしない形でしょうか？

**Comment by binnmti** [comment]
https://github.com/Greek-Academy/one-pager-maker/pull/103#pullrequestreview-2302800643
Change English!

File: .github/workflows/tag-deployment.yml [hunk 9]
```diff
@@ -0,0 +1,110 @@
+name: Tag-based Deployment
+
+on:
+  pull_request:
+    branches:
+      - main
+  # push:
+  #   tags:
+  #     - "v*"
+
```
[the remark this answers is not in the document]
Comment (binnmti) [reply .github/workflows/tag-deployment.yml]: revert

**Pull request merged**